### ISSUE TITLE
SUGGESTION for v5: Populate should avoid mutations

### ISSUE THREAD

**Issue by jlchereau** [Automattic/mongoose#5934]
IMHO, populate mutating an _id into a document is wrong. Mutations make building CRUD services more difficult than it should be because the object populated cannot be sent back with modifications without further processing to restore the id.

If possible, I would like to see an `alias` in v5 so that instead of populating the id, it would populate a new property named as defined in the `alias` below:

```
var storySchema = Schema({
  authorId: { type: Schema.Types.ObjectId, ref: 'Person', alias: 'author' },
  title: String,
});
```

This would be the equivalent of AS in SQL Queries.

Without designating an `alias`, the behavior would stay the same to remain compatible with previous versions.

**Comment by vkarpov15** [comment]
This is a very good suggestion, will consider this in more detail for a future minor release. In the meantime, virtual populate lets you do something similar with slightly different syntax:

```javascript
var storySchema = Schema({
  authorId: ObjectId,
  title: String
});

storySchema.virtual('author', {
  ref: 'Person',
  localField: 'authorId',
  foreignField: '_id',
  justOne: true
});
```

**Comment by jlchereau** [comment]
Unfortunately, I cannot progress with your suggestion because my model is actually

```
var storySchema = Schema({
  author: {
     _userId: ObjectId,
  }
  title: String
});
```

And what I am trying to achieve is populate the other 'author' properties inside the 'author' object based on _userId. In this case the virtual 'author' would compete with the actual 'author' in storySchema.

**Comment by vkarpov15** [comment]
Yeah that's a neat pattern. Right now I'd recommend just doing `authorId` and then an `author` virtual, because that makes it clear that `authorId` is always an ObjectId and `author` is always an object or `null`. Would that work as an alternative for you or do you have some very important reason for using this pattern?

**Comment by jlchereau** [comment]
Client side, my design includes observable models and collections and CRUD remote and local transport classes (with strategies for caching or synchronizing with a local database). A remote transport wraps ajax calls to a nodeJS + expressJS restful API. For each type of MongoDB  document, there is a full model to view/edit in a form and a lazy model with a limited set of fields for collections displayed in lists. For specific purposes there might be more variations.

The READ method of the remote transport class analyzes a full or lazy model to know the fields to request and the collection (or ajax endpoint) to request them from. To do so, the transport passes a fields param to the ajax request. **To keep this design  as generic as possible, the client-side model does not know which fields are native to the collection and which fields are populated**. Actually this design dates prior to the populate method in mongoose and I am trying to update it to use the latest mongoose features. 

In our example, a story could  be requested with an ajax call adding to the query string `&fields=title,author._userId,author.firstName,author.lastName`. The request is analyzed by an ExpressJS middleware which adds a mongoDB property to the req object. Then this mongoDB property is read in routes to make the MongoDB request using mongoose.

With the _"neat pattern"_ above, it is very easy for the route to discover which fields are native to the collection and which fields need to be populated. With your workaround I do not see immediately a generic way to relate author to authorId and populate an arbitrary number of fields passed by a request. but maybe I need to dig in a little bit more.

**Comment by vkarpov15** [comment]
Is it a major problem that your client can figure out which fields are populated and which are not? That doesn't sound too bad to me. You can also give the client the opportunity to specify whether they need the whole doc or just the ID. Also, you can `populate()` multiple virtuals the same way you `populate()` multiple normal fields, no difference between the two.

**Comment by wxs77577** [comment]
Hi @vkarpov15 

It's an important feature. The related fields in my mongodb are `*_id` (like, user_id, user_ids, course_id, etc.), and it's quite hard to match our ADMIN DASHBOARD if I use populate to modify `user_id` as an `User` object. I prefer to define the schema like this:
```
var storySchema = Schema({
  user_id: { type: Schema.Types.ObjectId, ref: 'Person', alias: 'author' },
  title: String,
});

Story.find().populate('author').exec()
```

out:
```
{
  "title": "Thank you.",
  "user_id": "a1343d234ffdeb",
  "user": {...}
}
```

I think it's better than a VIRTUAL type.

Thanks a lot.

**Comment by wxs77577** [comment]
Hi, @jlchereau  I've made some changes:

```
mongoose.Schema.prototype.relate = function (name, ref, localField, foreignField = '_id', justOne = true) {
  this.virtual(name, { ref, localField, foreignField, justOne })
}

//call
GroupSchema.relate('users', 'User', 'user_ids', '_id', false)

```

**Comment by jlchereau** [comment]
@vkarpov15, @wxs77577 
`it's quite hard to match our ADMIN DASHBOARD if I use populate to modify user_id as an User object` because mutation of data type is a real pain. Populating `user` alongside `user_id` based on a population alias that avoids mutations is the purpose of my suggestion and it is backward compatible.

**Comment by vkarpov15** [comment]
@wxs77577 I don't understand what you mean, an alias is just a special type of virtual.

@jlchereau I still don't quite understand why virtual populate doesn't solve your issue?

**Comment by jlchereau** [comment]
@vkarpov15 

Currently progressing with it:
- virtual populate does the job with a single embedded doc (justOne: true).
- I could not yet figure out something I am happy with when dealing with arrays of subdocuments.

**Comment by vkarpov15** [comment]
Whats the issue with justOne: false for arrays of subdocs?

**Comment by jlchereau** [comment]
@vkarpov15

The problem is not a bug in mongoose. It is about design decisions that make things more complicated than necessary.

Consider our framework as a sort of odata layer built on top of mongoose. We actually automatically translate any [Kendo UI query](https://docs.telerik.com/kendo-ui/api/javascript/data/datasource/methods/query) into a mongoose query.

This framework has a generic way of implementing things across collections and embedded arrays. This includes:
- CRUD operations - and this is were data mutations make everything more complicated, i.e. a GET/READ brings back a populated field that is an object, but a POST/CREATE sends the same field as a mongoose id used to populate the field: generally data mutations are BAD.
- lazy loading - when displaying items in a Kendo UI grid, our client framework automatically builds a list of fields to select based on the grid columns to reduce http payload. This requires a way to address each field in a request without knowing whether they are implemented as raw fields or virtually populated fields.

Virtual populate brings an acceptable solutions to such problem with justOne=true, provided you use a naming convention to somehow link the id field to the corresponding virtually populated object field:
- e.g., _userId is always an id and user is always a user object: no data mutation;
- @wxs77577 who commented above has the same problem but uses user_id.

Projection fields can therefore easily be addressed when building queries using 'user.firstName',  'user.lastName', ...: and the framework can work its magic : (1) user is a virtual populate? (2) YES, so populate with 'firstName', 'lastName', ... 

Note that I would have much preferred to have:
```
{
  ...
  user: {
    _id: xxxx
  }
  ...
}
as part of my raw object and to populate the other user fields as in
```
{
  ...
  user: {
    _id: xxxx,
    // populated fields
    firstName: 'John'
    lastName: 'Smith'
    ....
  }
  ...
}
```
which would avoid the ambiguity of having _userId and a possibly redundant populated user._id, or to have to hide completely _userId and only expose user._id in CRUD operations with translations based on the naming convention as and when required.

Bringing virtual populated arrays of subdocs with CRUD operations and lazy loading into this framework is significantly more complicated to achieve in a generic way but I am getting there.

I hope this makes the whole thread more clear.

**Comment by vkarpov15** [comment]
Yes that does make it more clear. How important is it to you - will you be able to work around this with virtual populate or are you blocked?
